### PR TITLE
Add a new run mode for retrying payments

### PR DESCRIPTION
**Pull request by amzid**
This PR resolves the issue #261. The following steps were performed:

* **Analysis**: The code in retry_failed.py needs to be adapted each time we change some other code parts. However, it can be integrated in main.py.

* **Solution**: We add a new run mode -M 4 to retry failed payments.

* **Performed tests**: Produce failed payment reports on Carthagenet and then use run mode 4 to pay them out. Payment reports are generated correctly.

* **Documentation**: NA

* **Check list**:

[ NA] I extended the Travis CI test units with the corresponding tests for this new feature (if needed).

[ NA] I extended the Sphinx documentation (if needed).

[NA ] I extended the help section (if needed).

[NA ] I changed the README file (if needed).

[ NA] I created a new issue if there is further work left to be done (if needed).

**Work effort**: 6 hrs.

### CODE REVIEW

File: src/pay/payment_producer.py [hunk 1]
```diff
@@ -100,7 +100,9 @@ def retry_fail_run(self):
     def run(self):
         # call first retry if not in onetime mode.
         # retry_failed script is more suitable for one time cases.
```
Comment (jdsika) [inline src/pay/payment_producer.py:102]: comment outdated

File: src/pay/payment_producer.py [hunk 2]
```diff
@@ -100,7 +100,9 @@ def retry_fail_run(self):
...
-        if not self.run_mode == RunMode.ONETIME:
+        if self.run_mode == RunMode.ONETIME:
```
Comment (jdsika) [inline src/pay/payment_producer.py:103]: I am confused that you remove the NOT case and add the true case?

**Review by jdsika — CHANGES_REQUESTED** [review]

**Comment by utdrmac** [comment]
My thoughts on this are to introduce another RunMode. RETRYFAIL 4. This mode would have the same behavior as the retry_failed script. Merging this with RunMode 3 changes the behavior. The retry_failed script did only that: it retried the failed txns then exited. If we merge this with M3, then it will retry failed txns, then continue processing payments for next cycle, then exit. This may not be desired behavior.

The difference between main.py and retry_failed.py is that main calls `p.start()` vs `p.retry_failed_payments(args.retry_injected)`

So it could be a simple change to main.py to check. Something like this:

```
if RunMode(args.run_mode) == RunMode.RETRYFAILED:
  p.retry_failed_payments(args.retry_injected)
else:
  p.start()
```

**Comment by amzid** [comment]
> Merging this with RunMode 3 changes the behavior. The retry_failed script did only that: it retried the failed txns then exited. If we merge this with M3, then it will retry failed txns, then continue processing payments for next cycle, then exit. 

Actually, it does not continue to do calculations for the next cycle. It retries the failed payments for the desired cycle, then since there is already one payment file for that cycle under the reports directory, it does not perform any further payments. 

> if RunMode(args.run_mode) == RunMode.RETRYFAILED:
>   p.retry_failed_payments(args.retry_injected)
> else:
>   p.start()
> ```
> 
> This way, none of the code in payment_producer needs to be modified.

But we still can change this way!

**Comment by amzid** [comment]
The use of the run mode 3 means the following: if no payment for the desired cycle was performed, this will be attempted. If a failed payment for the desired cycle was performed, this will be retried. If a successful payment for the desired cycle was performed, nothing will happen.
The other cycles will not be affected by any means.
I do not see the need for classifiying this case (If a failed payment for the desired cycle was performed, this will be retried.) under a fourth run mode.
However, if you think that adding a new run mode is more intuitive, we can change it.

**Comment by amzid** [comment]
I still need to adjust the code comments and block some console outputs.

**Comment by amzid** [comment]
> My thoughts on this are to introduce another RunMode. RETRYFAIL 4. This mode would have the same behavior as the retry_failed script. Merging this with RunMode 3 changes the behavior. The retry_failed script did only that: it retried the failed txns then exited. If we merge this with M3, then it will retry failed txns, then continue processing payments for next cycle, then exit. This may not be desired behavior.
> 
> The difference between main.py and retry_failed.py is that main calls `p.start()` vs `p.retry_failed_payments(args.retry_injected)`
> 
> So it could be a simple change to main.py to check. Something like this:
> 
> ```
> if RunMode(args.run_mode) == RunMode.RETRYFAILED:
>   p.retry_failed_payments(args.retry_injected)
> else:
>   p.start()
> ```
> 
> This way, none of the code in payment_producer needs to be modified.

I had a second look at how the retry_failed.py script behaves and I agree with you. There is no need to retry the failed payment for one single cycle, I will add a new run mode to retry all failed payments.

**Review by jdsika — APPROVED** [review]
OK for me, I would like @utdrmac to test it as well and make sure that the behaviour is correct on corner cases. 
1) Use without failed payments
2) Use with one
3) Use with multiple
4) Use but fail again

Check if all logging messaged and new file creations are ok

**Comment by jdsika** [comment]
@utdrmac I am waiting for your approval

**Comment by utdrmac** [comment]
@jdsika I added some comments. Waiting to hear back from @amzid 
......
ahh.. seems i never actually clicked the button to submit the comment. sorry.

File: src/pay/payment_producer.py [hunk 3]
```diff
@@ -98,94 +99,86 @@ def retry_fail_run(self):
...
-            pymnt_cycle = current_cycle - abs(self.initial_payment_cycle) - (self.nw_config['NB_FREEZE_CYCLE'] + 1)
-            logger.debug("Payment cycle is set to {}".format(pymnt_cycle))
-
-        while not self.exiting and self.life_cycle.is_running():
-
-            # take a breath
-            time.sleep(5)
+        if self.run_mode == RunMode.RETRY_FAILED:
+            self.retry_failed_payments(self.retry_injected)
+            time.sleep(10)
```
[the remark this answers is not in the document]
Comment (amzid) [reply src/pay/payment_producer.py:104]: That was my first implementation. But if you return here, the main function would run forever still the life cycle is running. The main thread is interrupted properly (=> set right flags and delete lock file) using the exit method. However, alternative suggestions to fix this would be welcome!

[the remark this answers is not in the document]
Comment (amzid) [reply src/pay/payment_producer.py]: Thanks for the suggestions. I implemented the recommended changes. Since we are now interrupting the main thread directly, the interrupt is visible in the log as shown in the image. I am not sure how this can be avoided by avoiding changes to payment_producer.py
image
I personally think that 635942d is more stable even if it includes changes to payment_producer.py.

**Review by utdrmac — APPROVED** [review]
